SI-3995 Exclude companions with an existential prefix.

retronym opened the pull request:
In `(qual: Q).apply(expr)` where `expr` must be implictily
converted to a path dependent type `T` defined in `qual`,
we were looking for companion implicits via a path prefixed
by an existential skolem `_1`. These aren't much good to us,
as when we try to feed them into `mkAttributedQualifer`, a
crash rightly ensues.

This commit excludes companions prefixed by an existentially
bound path.

Review by @adriaanm

retronym committed:
SI-3995 Exclude companions with an existential prefix.

In `(qual: Q).apply(expr)` where `expr` must be implictily
converted to a path dependent type `T` defined in `qual`,
we were looking for companion implicits via a path prefixed
by an existential skolem `_1`. These aren't much good to us,
as when we try to feed them into `mkAttributedQualifer`, a
crash rightly ensues.

This commit excludes companions prefixed by an existentially
bound path.

xeno-by commented on test/files/neg/t3995.scala:
```diff
@@ -0,0 +1,25 @@
+  // which are members of these companion objects."
+  //
+  // The skolem is stable, but it doen't seem much good to us
+  (new Lift).apply("")
```
Would `l.apply("")` work?

retronym replied:
Yeah, that's still fine. I'll add that to the test case.

xeno-by replied:
Crashes for me in unpatched 2.10.x

retronym replied:
Odd, can you post your full test? I've just pushed an expanded version of this test which allows `l.apply("")` and expands it to `Test.this.l.apply(Test.this.l.F0.f2f0(""));`

xeno-by replied:
My bad, it crashed with `scala.NotImplementedError: an implementation is missing`.

retronym replied:
Wait, you're trying to _run_ code to test it? :wink:

xeno-by replied:
I can't help myself. REPL is so convenient :)

retronym committed:
Expand test with a stably qualified example.

In the example below, we need a view from
`String => l.F0`, and the companion object `FO` is
reachable by a stable, non existentially-bound path.

    class Lift {
      def apply(f: F0) {}

      class F0
      object F0 {
        implicit def f2f0(fn: String): F0 = ???
      }
    }

    object Test {
      val l = new Lift
      l.apply("") // okay
    }

Followup for SI-3995

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1078/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1078/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1790/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1790/

retronym commented:
Paging @adriaanm, 

:musical_score: All I want for Christmas is a LGTM...

paulp commented:
LGTM. Ho ho ho.

adriaanm commented:
Thanks, Santa, but where's your beard?

I was fretting over how this corresponds to the spec. I have stopped worrying and it started LingGTM when I realized the candidates you exclude here would fail later anyway -- sometimes it helps to look at the bug report. 

Anyway, is there a more general criterion for when we can't stably refer to the qualifier? Could an existential be acceptable in certain cases?

paulp commented:
That is like word for word what I wrote, and then deleted for the terser "LGTM", after I decided that I'd do better to just go attempt to consolidate the various stability definitions.
